var/spack/repos/builtin/packages/py-cylc-flow/package.py

```diff
@@ -29,15 +32,32 @@ class PyCylcFlow(PythonPackage):
     depends_on("py-jinja2@3.0", type=("build", "run"))
     depends_on("py-metomi-isodatetime@3.0", type=("build", "run"), when="@:8.2.0")
     depends_on("py-metomi-isodatetime@3:3.1", type=("build", "run"), when="@8.2.3:")
-    depends_on("py-protobuf@4.21.2:4.21", type=("build", "run"))
+    depends_on("py-packaging", type=("build", "run"), when="@8.3:")
+    depends_on("py-protobuf@4.21.2:4.21", type=("build", "run"), when="@:8.2")
+    depends_on("py-protobuf@4.24.4:4.24", type=("build", "run"), when="@8.3:")
     depends_on("py-psutil@5.6.0:", type=("build", "run"))
     depends_on("py-pyzmq@22:", type=("build", "run"), when="@8.2:")
     depends_on("py-pyzmq@22", type=("build", "run"), when="@:8.1")
-    depends_on("py-importlib-metadata", type=("build", "run"), when="^python@:3.7")
-    depends_on("py-urwid@2", type=("build", "run"))
+    depends_on("py-importlib-metadata", type=("build", "run"), when="@:8.2 ^python@:3.7")
+    depends_on("py-importlib-metadata@5:", type=("build", "run"), when="@8.3: ^python@:3.11")
+    depends_on("py-urwid@2:2.6.1,2.6.4:2", type=("build", "run"))
     depends_on("py-rx", type=("build", "run"))
     depends_on("py-promise", type=("build", "run"))
     depends_on("py-tomli@2:", type=("build", "run"), when="^python@:3.10")
 
     # Non-Python dependencies
     depends_on("graphviz", type="run")
+
+    # cylc unsets PYTHONPATH, then sets it from CYLC_PYTHONPATH
```

Comment:
I know that this might be a controversial approach, because it means that the official documentation for cylc-flow@8.3.6 won't be correct for a spack install. But the alternative is to code up all dependency PYTHONPATH entries needed for py-cylc-flow manually in a setup_runtime_env routine and create a CYLC_PYTHONPATH for it. This may again clash with the documentation that suggests that the user has to set up CYLC_PYTHONPATH (which really doesn't make much sense with spack, where each Python dependency is installed in a separate directory).